File: spark-streaming/src/test/scala/com/example/spark/KafkaEventConsumerTest.scala
package com.example.spark

import org.scalatest.flatspec.AnyFlatSpec
import org.scalatest.matchers.should.Matchers
import com.example.protocol.EventProto.Event
import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.functions._

class KafkaEventConsumerTest extends AnyFlatSpec with Matchers {

  "Event deserialization" should "correctly parse Protobuf messages" in {
    val spark = SparkSession.builder()
      .appName("TestApp")
      .master("local[*]")
      .getOrCreate()

    try {
      import spark.implicits._

      // Create a test event
      val event = Event.newBuilder()
        .setId("test-123")
        .setName("TestEvent")
        .setTimestamp(System.currentTimeMillis())
        .setData("test data")
        .build()

      // Serialize to bytes
      val bytes = event.toByteArray

      // Create DataFrame with serialized data
      val df = Seq(bytes).toDF("value")

      // Apply deserialization UDF
      val deserializedDF = df.select(
        KafkaEventConsumer.deserializeEvent(col("value")).as("event")
      )

      // Collect results
      val result = deserializedDF.collect()

      result should have length 1
      val row = result.head
      val eventRow = row.getStruct(0)
      
      // Tuple field indices: 0=id, 1=name, 2=timestamp, 3=data
      eventRow.getAs[String](0) shouldBe "test-123"
      eventRow.getAs[String](1) shouldBe "TestEvent"
      eventRow.getAs[Long](2) shouldBe event.getTimestamp
      eventRow.getAs[String](3) shouldBe "test data"
    } finally {
      spark.stop()
    }
  }

  it should "handle null bytes gracefully" in {
    val spark = SparkSession.builder()
      .appName("TestApp")
      .master("local[*]")
      .getOrCreate()

    try {
      import spark.implicits._

      val df = Seq(null.asInstanceOf[Array[Byte]]).toDF("value")

      val deserializedDF = df.select(
        KafkaEventConsumer.deserializeEvent(col("value")).as("event")
      )

      val result = deserializedDF.collect()

      result should have length 1
      val row = result.head
      val eventRow = row.getStruct(0)
      
      // Tuple field indices: 0=id, 1=name, 2=timestamp, 3=data
      eventRow.getAs[String](0) shouldBe ""
      eventRow.getAs[String](1) shouldBe ""
      eventRow.getAs[Long](2) shouldBe 0L
      eventRow.getAs[String](3) shouldBe ""
    } finally {
      spark.stop()
    }
  }

  it should "deserialize multiple events" in {
    val spark = SparkSession.builder()
      .appName("TestApp")
      .master("local[*]")
      .getOrCreate()

    try {
      import spark.implicits._

      val events = (1 to 5).map { i =>
        Event.newBuilder()
          .setId(s"event-$i")
          .setName(s"Event$i")
          .setTimestamp(i.toLong)
          .setData(s"data-$i")
          .build()
          .toByteArray
      }

      val df = events.toDF("value")

      val deserializedDF = df.select(
        KafkaEventConsumer.deserializeEvent(col("value")).as("event")
      )

      val results = deserializedDF.collect()

      results should have length 5

      // Tuple field indices: 0=id, 1=name, 2=timestamp, 3=data
      results.zipWithIndex.foreach { case (row, idx) =>
        val i = idx + 1
        val eventRow = row.getStruct(0)
        eventRow.getAs[String](0) shouldBe s"event-$i"
        eventRow.getAs[String](1) shouldBe s"Event$i"
        eventRow.getAs[Long](2) shouldBe i.toLong
        eventRow.getAs[String](3) shouldBe s"data-$i"
      }
    } finally {
      spark.stop()
    }
  }
}

File: flink-streaming/src/test/java/com/example/flink/KafkaEventConsumerTest.java
package com.example.flink;

import com.example.protocol.EventProto.Event;
import org.junit.Test;

import java.io.IOException;

import static org.junit.Assert.*;

public class KafkaEventConsumerTest {

    @Test
    public void testEventDeserialization() throws IOException {
        // Create a test event
        Event event = Event.newBuilder()
                .setId("test-123")
                .setName("TestEvent")
                .setTimestamp(System.currentTimeMillis())
                .setData("test data")
                .build();

        // Serialize to bytes
        byte[] bytes = event.toByteArray();

        // Deserialize using the schema
        KafkaEventConsumer.EventDeserializationSchema schema =
                new KafkaEventConsumer.EventDeserializationSchema();
        Event deserializedEvent = schema.deserialize(bytes);

        // Verify
        assertNotNull(deserializedEvent);
        assertEquals("test-123", deserializedEvent.getId());
        assertEquals("TestEvent", deserializedEvent.getName());
        assertEquals(event.getTimestamp(), deserializedEvent.getTimestamp());
        assertEquals("test data", deserializedEvent.getData());
    }

    @Test
    public void testEventDeserializationWithEmptyData() throws IOException {
        Event event = Event.newBuilder()
                .setId("test-456")
                .setName("EmptyEvent")
                .setTimestamp(0L)
                .setData("")
                .build();

        byte[] bytes = event.toByteArray();

        KafkaEventConsumer.EventDeserializationSchema schema =
                new KafkaEventConsumer.EventDeserializationSchema();
        Event deserializedEvent = schema.deserialize(bytes);

        assertNotNull(deserializedEvent);
        assertEquals("test-456", deserializedEvent.getId());
        assertEquals("EmptyEvent", deserializedEvent.getName());
        assertEquals(0L, deserializedEvent.getTimestamp());
        assertEquals("", deserializedEvent.getData());
    }

    @Test
    public void testEventDeserializationWithSpecialChars() throws IOException {
        Event event = Event.newBuilder()
                .setId("test-789")
                .setName("SpecialEvent")
                .setTimestamp(1234567890L)
                .setData("Special chars: !@#$%^&*()")
                .build();

        byte[] bytes = event.toByteArray();

        KafkaEventConsumer.EventDeserializationSchema schema =
                new KafkaEventConsumer.EventDeserializationSchema();
        Event deserializedEvent = schema.deserialize(bytes);

        assertNotNull(deserializedEvent);
        assertEquals("test-789", deserializedEvent.getId());
        assertEquals("SpecialEvent", deserializedEvent.getName());
        assertEquals(1234567890L, deserializedEvent.getTimestamp());
        assertEquals("Special chars: !@#$%^&*()", deserializedEvent.getData());
    }

    @Test
    public void testMultipleEventsSerialization() throws IOException {
        KafkaEventConsumer.EventDeserializationSchema schema =
                new KafkaEventConsumer.EventDeserializationSchema();

        for (int i = 1; i <= 5; i++) {
            Event event = Event.newBuilder()
                    .setId("event-" + i)
                    .setName("Event" + i)
                    .setTimestamp((long) i)
                    .setData("data-" + i)
                    .build();

            byte[] bytes = event.toByteArray();
            Event deserializedEvent = schema.deserialize(bytes);

            assertNotNull(deserializedEvent);
            assertEquals("event-" + i, deserializedEvent.getId());
            assertEquals("Event" + i, deserializedEvent.getName());
            assertEquals((long) i, deserializedEvent.getTimestamp());
            assertEquals("data-" + i, deserializedEvent.getData());
        }
    }

    @Test(expected = IOException.class)
    public void testMalformedProtobufData() throws IOException {
        // Create malformed data (random bytes that aren't valid Protobuf)
        byte[] malformedBytes = new byte[]{1, 2, 3, 4, 5, 6, 7, 8, 9, 10};

        KafkaEventConsumer.EventDeserializationSchema schema =
                new KafkaEventConsumer.EventDeserializationSchema();
        
        // This should throw IOException due to invalid Protobuf data
        schema.deserialize(malformedBytes);
    }
}
